### PR TITLE
Allow actions to return a grammar's type parameter's associated type

### LINKED ISSUE #247: Grammars using generics' associated types generate code that fails to compile
# TLDR

Creating a grammar that is parameterized by a generic type that is bounded by a trait with associated types that are used as the return values of actions results in compilation failure of the generated code.

For example:

```rust
use std::str::FromStr;
use traits;

grammar<P>(callbacks: &mut P) where P: traits::ParseCallbacks;

pub Term: P::Term = {
    <n:Num> => n.into(),
    "(" <t:Term> ")" => t,
};

Num: P::Num = <s:r"[0-9]+"> => callbacks.number(i32::from_str(s).unwrap());
```

Where `traits::ParseCallbacks` is defined like this:

```rust
pub trait ParseCallbacks {
    type Term: From<Self::Num>;
    type Num;

    fn number(&mut self, i32) -> Self::Num;
}
```

## Steps to Reproduce

I've created a test case at https://github.com/fitzgen/lalrpop-generic-grammar-bug

```
$ git clone https://github.com/fitzgen/lalrpop-generic-grammar-bug.git`
$ cd lalrpop-generic-grammar-bug
$ cargo build
```

## Expected

Builds A-OK.

## Actual

Type errors in the generated parser code involving the grammar's generic type parameter's trait's associated types:

<details><pre><code>
$ cargo test
   Compiling lalrpop-generic-grammar v0.1.0 (file:///Users/fitzgen/src/lalrpop-generic-grammar)
error[E0433]: failed to resolve. Use of undeclared type or module `P`
  --> src/grammar.rs:16:15
   |
16 |         NtNum(P::Num),
   |               ^^^^^^ Use of undeclared type or module `P`

error[E0433]: failed to resolve. Use of undeclared type or module `P`
  --> src/grammar.rs:17:16
   |
17 |         NtTerm(P::Term),
   |                ^^^^^^^ Use of undeclared type or module `P`

error[E0433]: failed to resolve. Use of undeclared type or module `P`
  --> src/grammar.rs:18:20
   |
18 |         Nt____Term(P::Term),
   |                    ^^^^^^^ Use of undeclared type or module `P`

error[E0433]: failed to resolve. Use of undeclared type or module `P`
   --> src/grammar.rs:262:18
    |
262 |     ) -> (usize, P::Num, usize) {
    |                  ^^^^^^ Use of undeclared type or module `P`

error[E0433]: failed to resolve. Use of undeclared type or module `P`
   --> src/grammar.rs:272:18
    |
272 |     ) -> (usize, P::Term, usize) {
    |                  ^^^^^^^ Use of undeclared type or module `P`

error[E0433]: failed to resolve. Use of undeclared type or module `P`
   --> src/grammar.rs:282:18
    |
282 |     ) -> (usize, P::Term, usize) {
    |                  ^^^^^^^ Use of undeclared type or module `P`

error[E0433]: failed to resolve. Use of undeclared type or module `P`
  --> src/grammar.rs:16:15
   |
16 |         NtNum(P::Num),
   |               ^^^^^^ Use of undeclared type or module `P`

error[E0433]: failed to resolve. Use of undeclared type or module `P`
  --> src/grammar.rs:17:16
   |
17 |         NtTerm(P::Term),
   |                ^^^^^^^ Use of undeclared type or module `P`

error[E0433]: failed to resolve. Use of undeclared type or module `P`
  --> src/grammar.rs:18:20
   |
18 |         Nt____Term(P::Term),
   |                    ^^^^^^^ Use of undeclared type or module `P`

error[E0433]: failed to resolve. Use of undeclared type or module `P`
   --> src/grammar.rs:262:18
    |
262 |     ) -> (usize, P::Num, usize) {
    |                  ^^^^^^ Use of undeclared type or module `P`

error[E0433]: failed to resolve. Use of undeclared type or module `P`
   --> src/grammar.rs:272:18
    |
272 |     ) -> (usize, P::Term, usize) {
    |                  ^^^^^^^ Use of undeclared type or module `P`

error[E0433]: failed to resolve. Use of undeclared type or module `P`
   --> src/grammar.rs:282:18
    |
282 |     ) -> (usize, P::Term, usize) {
    |                  ^^^^^^^ Use of undeclared type or module `P`

error: aborting due to previous error(s)

error: Could not compile `lalrpop-generic-grammar`.
Build failed, waiting for other jobs to finish...
error: aborting due to previous error(s)

error: build failed
</code></pre></details>

The generated `__Symbol` contains associated types `P::Num` and `P::Term`, but isn't parameterized by the generic type `P` and is missing the `where P: traits::ParseCallbacks` clause as well:

https://github.com/fitzgen/lalrpop-generic-grammar-bug/blob/master/src/grammar.rs#L12-L19

```rust
pub enum __Symbol<'input> {
        Term_22_28_22(&'input str),
        Term_22_29_22(&'input str),
        Termr_23_22_5b0_2d9_5d_2b_22_23(&'input str),
        NtNum(P::Num),
        NtTerm(P::Term),
        Nt____Term(P::Term),
}
```

### PR DESCRIPTION
**Pull request by fitzgen**
Fixes #247

### CODE REVIEW

**Comment by fitzgen** [comment]
Ok this is working now!

Depends on #264

**Comment by fitzgen** [comment]
Ok, added more tests, as discussed on irc.

This is really ready for review now ;)

File: lalrpop/src/lr1/codegen/parse_table.rs [hunk 1]
```diff
@@ -311,10 +330,16 @@ impl<'ascent, 'grammar, W: Write> CodeGenerator<'ascent, 'grammar, W, TableDrive
         // if we are generating multiple parsers from the same file:
         rust!(self.out, "#[allow(dead_code)]");
         rust!(self.out,
-              "pub enum {}Symbol<{}> {{",
+              "pub enum {}Symbol<{}>",
```
Comment (nikomatsakis) [inline lalrpop/src/lr1/codegen/parse_table.rs]: I wonder... so currently we only include those lifetime/type-parameters that we *need* -- but maybe it'd be better, or at least easier, to include *all* of them, and just throw in a phantom data. Then we wouldn't have to scrape around so much to find out which ones are actually used, right?

Comment (fitzgen) [reply lalrpop/src/lr1/codegen/parse_table.rs]: Would all these symbols be in a a new, unused variant? That could have the unfortunate side effect of making the discriminate larger, right?

I guess that seems like a fine trade off if the finding of lifetime/type parameters is a code/maintenance burden and has a history of bugs. If not, then it doesn't seem worth it. What do you think?

Comment (nikomatsakis) [reply lalrpop/src/lr1/codegen/parse_table.rs]: Yes, it'd make the discriminant larger. That seems like a non-issue to me. But let's just merge this PR for now, I think, and we can revisit later.

**Pull request merged**